### PR TITLE
feat: use state machines!

### LINKED ISSUE #3: Connect the dismiss button to the close action
Right now clicking "dismiss" doesn't actually do anything. We should fix that. 😄

### PR DESCRIPTION
**Pull request by jlengstorf**
On the _Learn With Jason_ livestream today, @davidkpiano and I refactored the feedback widget to use xstate.

[Watch us refactor the code live.](https://www.youtube.com/watch?list=PLz8Iz-Fnk_eTpvd49Sa77NiF8Uqq5Iykx&v=czi24DqUfSA)

I also refactored the code for readability a bit:
- Moved views to their own component files
- Moved shared components to their own files

closes #3

### CODE REVIEW

File: src/components/feedback-widget/feedback-widget.js [hunk 1]
```diff
@@ -1,89 +1,82 @@
...
+        SUBMIT: "submitting",
+      },
+    },
+    submitting: {
+      invoke: {
+        id: "postFeedback",
+        src: ctx => postFeedback(ctx),
+        onDone: "success",
+        onError: "failed",
+      },
```
Comment (davidkpiano) [inline src/components/feedback-widget/feedback-widget.js:58]: Nice, we didn't even go over `invoke` but you nailed it here!

File: src/components/feedback-widget/feedback-widget.js [hunk 2]
```diff
@@ -1,89 +1,82 @@
...
+    },
+    success: {
+      // TODO can we send these events to gatsby-telemetry?
+      onEntry: "focusSuccessTitle",
```
Comment (davidkpiano) [inline src/components/feedback-widget/feedback-widget.js:62]: You can do that declaratively by using multiple actions:

```js
onEntry: ['focusSuccessTitle', 'logTelemetry'],
```

File: src/components/feedback-widget/feedback-widget.js (hunk 1 above)
Comment (jlengstorf) [reply src/components/feedback-widget/feedback-widget.js:58]: The docs were super helpful here. 💪💜

File: src/components/feedback-widget/feedback-form.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
+import React from "react"
+import { keyframes } from "@emotion/core"
+import styled from "@emotion/styled"
+import WidgetWrapper from "./widget-wrapper"
+import { SubmitButton, CloseButton } from "./buttons"
```
Comment (marcysutton) [inline src/components/feedback-widget/feedback-form.js:5]: Nice move into a single file! 👍

**Review by marcysutton — APPROVED** [review]
This looks fantastic! I added a commit to suppress scripted focus outlines and reduce the animation if the user requests it (using prefers-reduced-motion). Should be good to go!

File: src/components/feedback-widget/feedback-widget.js [hunk 4]
```diff
@@ -1,89 +1,81 @@
...
+  context: {
+    rating: 2,
+    comment: "",
+  },
+  states: {
+    opened: {
+      onEntry: "focusWidgetTitle",
+      on: {
+        RATE: {
+          actions: assign({ rating: (_ctx, e) => e.value }),
```
Comment (marcysutton) [inline src/components/feedback-widget/feedback-widget.js]: What is `ctx`? I see it in the Xstate docs, but it's not super readable/intuitive for people new to it...

Comment (davidkpiano) [reply src/components/feedback-widget/feedback-widget.js]: Would this make it more clear?

```suggestion
          // Assign the RATE event value to the rating context
          actions: assign({ rating: (_context, event) => event.value }),
```

Comment (marcysutton) [reply src/components/feedback-widget/feedback-widget.js]: Ooh yes, this is much better!

Comment (davidkpiano) [reply src/components/feedback-widget/feedback-widget.js]: Thanks for helping me realize the abbreviation can be confusing. Made the change docs-wide here: https://github.com/davidkpiano/xstate/commit/1d3215d2b5408e95ad79a51793d5d276275ce80f

File: src/components/feedback-widget/feedback-form.js [hunk 5]
```diff
@@ -43,6 +43,12 @@ const Form = styled("form")`
       animation-delay: -500ms;
     }
   }
+  @media screen and (prefers-reduced-motion: reduce) {
```
Comment (jlengstorf) [inline src/components/feedback-widget/feedback-form.js:46]: Whoa, I didn’t know this existed. So cool!

**Pull request merged**